ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,670 @@
+					}
+				}
+			}
+		}
+	}
+
+	/// Construct an new state from given checkpoint header.
+	fn new_checkpoint_state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
+		assert_eq!(header.number() % self.epoch_length, 0);
+
+		let mut state = CliqueBlockState::new(
+			match header.number() {
+				0 => NULL_AUTHOR,
+				_ => recover_creator(header)?,
+			},
+			extract_signers(header)?);
+
+		state.calc_next_timestamp(header, self.period);
+
+		Ok(state)
+	}
+
+	fn state_no_backfill(&self, hash: &H256) -> Option<CliqueBlockState> {
+		return self.block_state_by_hash.write().get_mut(hash).cloned();
+	}
+
+	/// get CliqueBlockState for given header, backfill from last checkpoint if needed.
+	fn state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
+		let mut block_state_by_hash = self.block_state_by_hash.write();
+		if let Some(state) = block_state_by_hash.get_mut(&header.hash()) {
+			return Ok(state.clone());
+		}
+		// If we are looking for an checkpoint block state, we can directly reconstruct it.
+		if header.number() % self.epoch_length == 0 {
+			let state = self.new_checkpoint_state(header)?;
+			block_state_by_hash.insert(header.hash().clone(), state.clone());
+			return Ok(state);
+		}
+		// BlockState is not found in memory, which means we need to reconstruct state from last checkpoint.
+		match self.client.read().as_ref().and_then(|w| { w.upgrade() }) {
+			None => {
+				return Err(From::from("failed to upgrade client reference"));
+			}
+			Some(c) => {
+				let last_checkpoint_number = (header.number() / self.epoch_length as u64) * self.epoch_length;
+				assert_ne!(last_checkpoint_number, header.number());
+
+				let mut chain: &mut VecDeque<Header> = &mut VecDeque::with_capacity(
+					(header.number() - last_checkpoint_number + 1) as usize);
+
+				// Put ourselves in.
+				chain.push_front(header.clone());
+
+				// populate chain to last checkpoint
+				let mut last = chain.front().ok_or(
+					"just pushed to front, reference must exist;")?.clone();
+
+				while last.number() != last_checkpoint_number + 1 {
+					match c.block_header(BlockId::Hash(*last.parent_hash())) {
+						None => {
+							return Err(Box::new(
+								format!("parent block {} of {} could not be recovered.",
+										&last.parent_hash(), &last.hash())
+							).into());
+						}
+						Some(next) => {
+							chain.push_front(next.decode()?.clone());
+							last = chain.front().ok_or(
+								"just pushed to front, reference must exist;")?.clone();
+						}
+					}
+				}
+
+				// Catching up state, note that we don't really store block state for intermediary blocks,
+				// for speed.
+				let backfill_start = time::Instant::now();
+				info!(target: "engine",
+						"Back-filling block state. last_checkpoint_number: {}, target: {}({}).",
+						last_checkpoint_number, header.number(), header.hash());
+
+				// Get the state for last checkpoint.
+				let last_checkpoint_hash = *(chain.front().ok_or(
```

Comment:
should be fine to use `expect`  instead of `ok_or` because the `VecDeque` will have a least one element bacause we have inserted the `header`